dos/src/main.rs

```diff
@@ -23,14 +33,128 @@ fn get_repair_contact(nodes: &[ContactInfo]) -> ContactInfo {
     contact
 }
 
+/// Options for data_type=transaction
+#[derive(Serialize, Deserialize, Debug)]
+struct TransactionParams {
+    unique_transactions: bool, // use unique transactions
+    num_sign: usize,           // number of signatures in a transaction
+    valid_block_hash: bool,    // use valid blockhash or random
+    valid_signatures: bool,    // use valid signatures or not
+    with_payer: bool,          // provide a valid payer
+}
+
+struct TransactionGenerator {
+    blockhash: Hash,
+    last_generated: Instant,
+    transaction_params: TransactionParams,
+    cached_transaction: Option<Transaction>,
+}
+
+impl TransactionGenerator {
+    fn new(transaction_params: TransactionParams) -> Self {
+        TransactionGenerator {
+            blockhash: Hash::default(),
+            last_generated: (Instant::now() - Duration::from_secs(100)),
+            transaction_params,
+            cached_transaction: None,
+        }
+    }
+
+    fn generate(&mut self, payer: &Keypair, rpc_client: &Option<RpcClient>) -> Transaction {
+        if !self.transaction_params.unique_transactions && self.cached_transaction != None {
+            return self.cached_transaction.as_ref().unwrap().clone();
+        }
+
+        // generate a new blockhash every 1sec
+        if self.transaction_params.valid_block_hash
+            && self.last_generated.elapsed().as_millis() > 1000
+        {
+            self.blockhash = rpc_client.as_ref().unwrap().get_latest_blockhash().unwrap();
+            self.last_generated = Instant::now();
+        }
+
+        let lamports = 5;
+        let transfer_instruction = SystemInstruction::Transfer { lamports };
+        let program_ids = vec![system_program::id(), stake::program::id()];
```

Comment:
Could you add some documentation about what these program_ids are for? ie. that this tool constructs a transfer from the System program to the Stake program, and why

Reply:
The idea was to have some valid instructions and I have picked up some arbitrary valid instructions from [test_tx.rs](https://github.com/solana-labs/solana/blame/c9c78622a842ea795c7d37dc0f1b42b9e9ce3e20/perf/src/test_tx.rs#L29).

Reply:
Sure. I should probably mention that even if transaction has payer and valid blockhash, it will fail anyways during the check for instruction signatures: https://github.com/solana-labs/solana/blob/master/runtime/src/system_instruction_processor.rs#L237

So would be nice to know if there is a way to make it pass this check

Reply:
You mean you want it to enter the instruction processor and pass the signature check, but fail processing?
Seems like if you just create a new (unfunded) Keypair to be the `from` account (index[0] in the Transfer ix) and have that sign, it should do what you want...? Or have I missed what you're trying to do?